### PR TITLE
Add MachineProvisionImagePullPolicy setting

### PR DESCRIPTION
**Pull request by felipe-colussi**
# This will not be merged. Waiting for  release v2.8! 
I just opened the PR now to get reviews / opinions on the code.

## Problem

A client needed have an air-gapped environment and was having problems wile creating a new cluster. That happens due to the `ImagePullPolicy`  of the `...-machine-provisioning` job being set to `Always`.

## Solution

As discussed on the jira ticket the idea is to enable the client to set the image pull policy through global settings. 
If this is not configured the behavior will remain the same, that's why the default value is Always. 
The setting will be enabled in initialization (once set it cannot be changed unless rancher is restarted).

## Some questions: 

- [x] Are the function / variable names ok ? 
- [x] Should GetMachineProvisionImagePullPolicy() exist? I did it to avoid errors if the user set an invalid ImagePullPolicy.  
- [x] GetMachineProvisionImagePullPolicy() Should be on settings.go? I left it there because there are other "getters" on the file. 
- [x] If the user set the ImagePullPolicy to Never without the image the pod will get stuck in ErrImageNeverPull, do we allow "Never" as a valid input? 
 
### Manual Testing

Running rancher locally, the process was basically:
*  set the  env `CATTLE_MACHINE_PROVISION_IMAGE_PULL_POLICY`
*  start rancher/rancher
*  create a downstream cluster using the UI.
*  `kubectl edit job {JOB-NAME} -n fleet-default` and check the ImagePullPolicy.
*  `kubectl describe pod {POD-NAME} -n fleet-default` and check if the image was pulled or respected the value set. 

I did that test with `CATTLE_MACHINE_PROVISION_IMAGE_PULL_POLICY`  set to:  `""` (empty), `Always`, `Never`, `IfNotPresent`, `wrong` and without it set. 

### Automated Testing

* Test types added/modified:
    * Unit
    * None
    If the GetMachineProvisionImagePullPolicy is Ok i'll do unit test for it. Just waiting to confirm if it is ok.

### CODE REVIEW

**Review by a-blender — COMMENTED** [review]
Lgtm in general,

* Are the function / variable names ok? - yes
* Should GetMachineProvisionImagePullPolicy() exist? I did it to avoid errors if the user set an invalid ImagePullPolicy. - yes looks fine
* GetMachineProvisionImagePullPolicy() should be on settings.go? I left it there because there are other "getters" on the file. - I think so yeah, it's scoped to the env variable in settings.go so I'd keep it in that file.
* If the user set the ImagePullPolicy to Never without the image the pod will get stuck in ErrImageNeverPull, do we allow "Never" as a valid input? - we should. Any customer with an air gapped env should be able to set `Never` so that Rancher doesn't try to pull pod images and fail.

Since this is a global setting I do have concerns around a user configuring this. Will the images being pulled always be in the local cluster? If it's downstream, can a user have both an air gapped cluster and a non-air gapped cluster? This would cause issues on one cluster since the `ImagePullPolicy` is not per cluster.

**Comment by felipe-colussi** [comment]
> Since this is a global setting I do have concerns around a user configuring this. Will the images being pulled always be in the local cluster? If it's downstream, can a user have both an air gapped and a non-air gapped cluster? This would cause issues on one cluster since `ImagePullPolicy` is not per cluster.

The `MachineProvisionImage` and therefore the `MachineProvisionImagePullPolicy` settings are just used on the "local cluster" wile creating a downstream clusters.  That was the reason for setting it on a "global" config. 

Related to your comment, I also noticed that  we have [another case](https://github.com/rancher/rancher/blob/39cbdbd706c0bc75a3576f1b1a6a87bd3ad87d4a/pkg/controllers/management/usercontrollers/controller.go#L290) that "Always" is set as the default pull policy, this case happens on the downstream-cluster, wile deleting them. We are waiting for confirmation, but it will not be part of the scope of this issue.

File: pkg/settings/setting.go
Comment (HarrisonWAffel) [inline pkg/settings/setting.go:455]: This comment isn't formatted properly. We should remove the indentation and white space between the two sentences

File: pkg/settings/setting.go
Comment (HarrisonWAffel) [inline pkg/settings/setting.go:462]: While allowing `v1.PullNever` feels like a foot gun, I think it's a good idea to allow the user to set any of the available pull options as they see fit. Plus, in air-gapped scenarios where a registry has not been configured, `PullIfNotPresent` and `Never` will be functionally the same.

**Review by HarrisonWAffel — CHANGES_REQUESTED** [review]
Generally looks good, but we should fix the function comment format before we merge this. Also, feel free to write unit tests even if we aren't confident in some of the naming conventions, it can always be refactored with the new name. 

Also, imo we can remove the regression concern, since the default value used in `GetMachineProvisionImagePullPolicy` equals the default value in current versions of Rancher. Because of this, your PR will not result a behavior change by default(which is good!). It does create room for misconfiguration of Rancher, but that is different than a regression. 

I think it's important to note that this problem relates to a relatively unusual air-gap setup, where a dedicated registry is not used and all images are simply stored on the same node as Rancher.

**Review by a-blender — COMMENTED** [review]
@felipe-colussi Thanks for clarifying.

* What about this case as well? https://github.com/rancher/rancher/blob/52496c9026b46dd3588dfa680552d4b6ee52ec9b/pkg/controllers/capr/machineprovision/args.go#L164 @jakefhyde does the capr controller run in the local cluster? Should the `ImagePullPolicy` be configurable here for provisioned machines?
* Can we add any automated tests for this?

File: pkg/settings/setting.go
Comment (felipe-colussi) [reply pkg/settings/setting.go]: Done

File: pkg/settings/setting_test.go
Comment (a-blender) [inline pkg/settings/setting_test.go:91]: For tests, I use the `input/expected` format for test cases. This is also throughout the Rancher code. Here's a good example https://github.com/rancher/rancher/blob/59b00c6f9a0a7c598e24c491c045a0fc0586ba6b/pkg/catalogv2/system/system_test.go#L34-L50.

**Review by a-blender — CHANGES_REQUESTED** [review]

File: pkg/settings/setting.go
Comment (snasovich) [inline pkg/settings/setting.go:456]: If this is the cleanest way to check if the passed value is one of supported values then I'm just going to rant that Go needs better support for enumerations. :(

File: pkg/settings/setting.go
Comment (snasovich) [inline pkg/settings/setting.go:464]: Let's change it to Warn, it's not an error

File: pkg/settings/setting_test.go
Comment (snasovich) [inline pkg/settings/setting_test.go:110]: Could we please also add test case for handling empty value?

**Review by snasovich — COMMENTED** [review]
Couple minor comments to address + one rant that can probably be ignored.
Doing "Comment" instead of "Request Changes" to not block merging if I can't get back to it.

**Review by HarrisonWAffel — APPROVED** [review]

File: pkg/settings/setting_test.go
Comment (HarrisonWAffel) [inline pkg/settings/setting_test.go:6]: We try to separate imports from the standard library and other imports with a newline, 
i.e.
```go
import (
	"encoding/json"
	"fmt"	
	"io"
	"os"
	"testing"

	v1 "k8s.io/api/core/v1"
        "github.com/sirupsen/logrus"
)
```
This is a pattern we try and follow throughout Rancher

**Review by HarrisonWAffel — CHANGES_REQUESTED** [review]
small nit

File: pkg/settings/setting_test.go
Comment (felipe-colussi) [reply pkg/settings/setting_test.go]: Done.  
Is there a reason why you putted the v1 over logrus?

**Review by HarrisonWAffel — APPROVED** [review]

File: pkg/settings/setting_test.go
Comment (HarrisonWAffel) [reply pkg/settings/setting_test.go]: No, I just wrote the code block quickly and didn't sort the names properly

**Review by jiaqiluo — APPROVED** [review]
LGTM

**Pull request merged**
